File: 정주완/Programmers/Level 3/표현 가능한 이진트리.py
# 45m

def is_possible(string):
    if len(string) not in [1,3,7,15,31,63]:
        return is_possible('0' + string)

    root_node = len(string)//2

    # 루트 노드가 더미 노드(0)인 경우
    if string[root_node] == '0':
        # 자식들도 모두 더미(0)인 경우는 True
        if int(string) == 0:
            return True
        # 자식 중 더미 노드가 아닌 노드(1)가 있으면 False
        else:
            return False

    # 루트 노드가 더미가 아니고, 더 나눌 수 없는 경우는 True
    if len(string) == 1 or len(string) == 3:
        return True      

    # 왼쪽 서브트리와 오른쪽 서브트리에 대해 검사한 결과를 합친다.
    return is_possible(string[0:root_node]) and is_possible(string[root_node + 1:])


def solution(numbers):
    answer = []    
        
    for number in numbers:
        binary_number_string = str(format(number,'b'))
        if is_possible(binary_number_string):
            answer.append(1)
        else:            
            answer.append(0)
    return answer

File: 정주완/Programmers/Level 3/표현 가능한 이진트리.java
import java.util.Arrays;
import java.util.List;

class Solution {

    // 이진트리로 표현 가능한지 재귀적으로 확인하는 함수
    public static boolean isPossible(String binaryString) {
        int length = binaryString.length();

        // 길이가 1, 3, 7, 15, 31, 63이 아니면 앞에 '0' 추가
        List<Integer> validLengths = Arrays.asList(1, 3, 7, 15, 31, 63);
        if (!validLengths.contains(length)) {
            return isPossible("0" + binaryString);
        }

        // 루트 노드 확인 (문자열의 중간 값)
        int rootNode = length / 2;

        // 루트 노드가 더미 노드(0)인 경우
        if (binaryString.charAt(rootNode) == '0') {
            // 자식들도 모두 더미 노드(0)일 경우
            if (binaryString.chars().allMatch(ch -> ch == '0')) {
                return true;
            }
            // 자식 중 더미가 아닌 노드(1)가 있을 경우
            return false;
        }

        // 더 이상 나눌 수 없는 경우
        if (length == 1 || length == 3) {
            return true;
        }

        // 좌우 서브트리를 재귀적으로 검사
        boolean left = isPossible(binaryString.substring(0, rootNode));
        boolean right = isPossible(binaryString.substring(rootNode + 1));

        return left && right;
    }

    public int[] solution(long[] numbers) {
        int[] answer = new int[numbers.length];

        for (int i = 0; i < numbers.length; i++) {
            String binaryString = Long.toBinaryString(numbers[i]);

            if (isPossible(binaryString)) {
                answer[i] = 1;
            } else {
                answer[i] = 0;
            }
        }

        return answer;
    }
}
